File: packages/python/peru_spatial_utils/territorial/calculate_territorial_risk.py
from typing import Iterable, Mapping, TypedDict

_WEIGHTS = {"BAJO": 10, "MEDIO": 20, "ALTO": 34}


class TerritorialRisk(TypedDict):
    risk: str
    score: int


def calculate_territorial_risk(
    intersections: Iterable[Mapping[str, str]],
) -> TerritorialRisk:
    total = sum(_WEIGHTS.get(item.get("level", ""), 0) for item in intersections)
    score = min(100, total)
    if score >= 67:
        risk = "ALTO"
    elif score >= 34:
        risk = "MEDIO"
    else:
        risk = "BAJO"
    return {"risk": risk, "score": score}

File: packages/rust/src/territorial/calculate_territorial_risk.rs
use super::types::{Intersection, TerritorialRisk};

pub fn calculate_territorial_risk(intersections: &[Intersection]) -> TerritorialRisk {
    let sum: i32 = intersections.iter().map(|i| weight(&i.level)).sum();
    let score = sum.min(100);
    let risk = if score >= 67 {
        "ALTO"
    } else if score >= 34 {
        "MEDIO"
    } else {
        "BAJO"
    };
    TerritorialRisk {
        risk: risk.to_string(),
        score,
    }
}

fn weight(level: &str) -> i32 {
    match level {
        "BAJO" => 10,
        "MEDIO" => 20,
        "ALTO" => 34,
        _ => 0,
    }
}
